\documentclass[webpdf,mynatbib,nosurname,nogrid,noCE,noMSC]{SYS}

\usepackage{cpl-sft}
\usepackage{url}
\usepackage{graphicx}
\usepackage{subfigure}
\usepackage{enumitem}

\setcounter{page}{1}

\pvtitle{Software for Systematics and Evolution}
\volname{}
\jvolume{66}
\jshort{syw}
\jvol{VOL. 66}
\jissue{2}
\cename{PR}

\mstype{Software for Systematics and Evolution}

\pubyear{2018}

\artid{087}

\newcommand{\tracer}{\emph{Tracer}}




\begin{document}

\title[RAMBAUT ET~AL.---TRACER 1.7]{
Posterior summarisation in Bayesian phylogenetics using Tracer 1.7
}

\author[]{
\textsc{Andrew} \surname{\textsc{Rambaut}}\,$^{1}$,
\textsc{Alexei J.} \surname{\textsc{Drummond}}\,$^{2,3}$,
\textsc{Dong} \surname{\textsc{Xie}}\,$^{2,3}$,
\textsc{Guy} \surname{\textsc{Baele}}\,$^{4}$,
\textsc{Marc A.} \surname{\textsc{Suchard}}\,$^{5,6}$
}



\address{
$^{1}$Institute of Evolutionary Biology, University of Edinburgh, Edinburgh, UK;
$^{2}$Department of Computer Science, University of Auckland, Auckland, NZ;
$^{3}$Centre for Computational Evolution, University of Auckland, Auckland, NZ;
$^{4}$Department of Microbiology and Immunology, Rega Institute, KU Leuven - University of Leuven, Leuven, Belgium;
$^{5}$Department of Human Genetics, University of California, Los Angeles, USA;
$^{6}$Department of Biostatistics, University of California, Los Angeles, USA \\[6pt]
$^{\ast}$Correspondence to:
\href{a.rambaut@ed.ac.uk}{\url{a.rambaut@ed.ac.uk}},
\href{alexei@cs.auckland.ac.nz}{\url{alexei@cs.auckland.ac.nz}}
and
\href{msuchard@ucla.edu}{\url{msuchard@ucla.edu}}}


\history{Received X February 2018; reviews returned X March 2018; accepted X April 2018}

\editor{Associate Editor: Edward Susko}

\abstract{
Bayesian inference of phylogeny using Markov chain Monte Carlo (MCMC) plays a central role in understanding evolutionary history from molecular sequence data.
Visualising and analysing the MCMC-generated samples from the posterior distribution is a key step in any non-trivial Bayesian inference.
We present the software package \tracer\ (version 1.7) for visualising and analysing the MCMC trace files generated through Bayesian phylogenetic inference.
\tracer\ provides kernel density estimation, multivariate visualisation, demographic trajectory reconstruction, conditional posterior distribution summary and more.
\tracer\ is open-source and available at \url{http://beast.community/tracer}.
[Bayesian inference; phylogenetics; Markov chain Monte Carlo; visualisation]
\vspace*{-25pt}
}

\maketitle

\indent Bayesian inference of phylogeny using Markov chain Monte Carlo (MCMC) \citep{rannala1996probability, mau1999bayesian, drummond2002estimating} flourishes as a popular approach to uncover the evolutionary relationships among taxa, such as genes, genomes, individuals or species.
MCMC approaches generate samples of model parameter values - including the phylogenetic tree - drawn from their posterior distribution given molecular sequence data and a selection of evolutionary models.
Visualising, tabulating and marginalising these samples is critical for approximating the posterior quantities of interest that one reports as the outcome of a Bayesian phylogenetic analysis.
To facilitate this task, we have developed the \tracer\ (version 1.7) software package to
process MCMC trace files containing parameter samples and to interactively explore the high-dimensional posterior distribution.
\tracer\ works automatically with sample output from BEAST \citep{drummond2012bayesian}, BEAST2 \citep{bouckaert2014beast2},  LAMARC \citep{kuhner2006lamarc},  Migrate \citep{beerli2006comparison}, MrBayes \citep{ronquist2012mrbayes}, RevBayes \citep{hohna2016revbayes} and possibly other MCMC programs from other domains.

\vspace{-2em}
\section*{\textsc{Design and Implementation}}

\tracer\ examines the posterior samples from all the available parameters - treating continuous, integer and categorical parameters appropriately - from a trace and presents statistical summaries and visualisations.
Further, \tracer\ can analyse a single trace or combine samples from multiple files.
Immediately apparent in the default \tracer\ view, the effective sample size (ESS) is one such statistic that
allows users to assess the number of effectively independent draws from the posterior distribution the trace represents (Figure \ref{fig:overview}a).
Colour coding assists the user in determining potential MCMC mixing problems, with arbitrary cut-off values at 100 and 200.

Selecting multiple parameters from the ``Traces" panel on the left generates a side-by-side comparison or an overlay of the selected parameters' visualisations (Figure \ref{fig:overview}b-e).
Multiple trace files can be selected in a similar fashion to compare posterior samples between different replicates of an analysis.
If multiple trace files contain the same collection of parameters,  then a ``Combined'' trace appears automatically.
\tracer\ generates four display panels for the selected parameters:

\begin{itemize}[leftmargin=*]
\vspace{-0.5em}
\item Estimates: Reports common summary statistics such as the sample mean, standard deviation, highest posterior density interval and ESS.  Also presents a histogram of sample values for a single selected parameter (Figure \ref{fig:overview}a) or  side-by-side boxplots for multiple continuous parameters (Figure \ref{fig:overview}b).
\item Marginal Density: Draws density plots for the selected parameter(s), including  kernel density estimates (Figure \ref{fig:overview}c), histograms and violin plots (Figure \ref{fig:overview}d) for continuous parameters and frequency plots for categorical or integer parameters.
\vspace*{-0.25em}
\item Joint-Marginal: Visualisation in this panel appears after selecting two or more parameters and the plot form depends on the parameter types. We show several examples in the next section of the paper.
%
\begin{figure}[t]
\subfigure[Tracer overview]{\includegraphics[width=0.485\textwidth]{figure_1a_rabv-overview.pdf}} \\
\subfigure[Boxplot]{\includegraphics[width = .23\textwidth, height = 3cm]{figure_1b_rabv-boxplots.pdf}}
\subfigure[Density]{\includegraphics[width = .23\textwidth, height = 3cm]{figure_1c_rabv-density.pdf}}\\
\subfigure[Violin]{\includegraphics[width = .23\textwidth, height = 3cm]{figure_1d_rabv-violin.pdf}}
\subfigure[Trace]{\includegraphics[width = .23\textwidth, height = 3cm]{figure_1e_rabv-traces.pdf}}
\vspace{-0.25cm}
\caption{Overview of \tracer\ functionality and individual parameter visualisations: (a) Main \tracer\ panel upon loading a single trace file; (b) boxplot representation of two continuous parameters; corresponding (c) kernel density estimates; (d) violin plots; and (e) the actual traces connecting the parameter values visited by the Markov chain.}
\label{fig:overview}
\end{figure}
%

%
\begin{figure}[t]
\subfigure[Bubble chart]{\includegraphics[width = .23\textwidth, height = 3cm]{figure_2a_rabv-integerjointmarginal.pdf}}
\subfigure[Marginal density]{\includegraphics[width = .23\textwidth, height = 3cm]{figure_2b_rabv-integermarginaldensity.pdf}} \\
\subfigure[Scatter plot]{\includegraphics[width = .23\textwidth, height = 3cm]{figure_2c_rabv-correlation.pdf}}
\subfigure[Correlation plot]{\includegraphics[width = .23\textwidth, height = 3cm]{figure_2d_rabv-joint-real.pdf}}
\vspace{-0.25cm}
\caption{Multi-parameter visualisations of: (a) the joint probability distribution of two integer variables through a bubble chart; (b) the marginal density of multiple integer or categorical variables through frequency plots; (c) two continuous variables through a classic scatter or correlation plot; (d) multiple ($> 2$) continuous variables using large correlation matrices.
}
\label{fig:4tabs}
\end{figure}
%
\item Trace: Constructs line plots connecting the sequential samples of one or more selected parameters against state or generation number (Figure \ref{fig:overview}e).  Users typically use this plot to assess mixing, select a suitable burn-in and identify trends that suggest convergence issues.
\end{itemize}
%

\tracer\ offers a solution of visualising conditional posterior distributions as well.
Selecting one continuous and one integer or categorical parameter generates side-by-side violin or boxplots under the Joint-Marginal panel.
These plots present the continuous parameter distribution conditioned on the unique integer or categorical values.
A typical use case involves Bayesian stochastic search variable selection (BSSVS), a form of model averaging, in which parameters influence the likelihood function only when a specific model is selected by a random indicator function.
Under BSSVS, a posterior estimate of the parameter should only sample values from states where the indicator equals one.
Discrete phylogeographic analyses frequently employ BSSVS due to the potentially large amount of transition rates that need to be estimated \citep{Lemey2009}, but this is also relevant when employing model averaging approaches, e.g., over relaxed molecular clocks \citep{Li2012}.

Finally, \tracer\ provides demographic reconstruction resulting in a graphical plot, often applied to reconstruct epidemic dynamics.
Available models include, e.g., constant size, exponential and logistic growth \citep{drummond2002estimating},
and the non-parametric Bayesian skyline \citep{drummond2005bayesian,heledDrummond2008}, skyride \citep{minin2008smooth} and skygrid \citep{gill2012improving}.

\vspace{-0.7cm}

\section*{\textsc{Example}}

\subsection*{Cross-species dynamics of North American bat rabies}

We use \tracer\ to infer the spatial dispersal and cross-species dynamics of rabies virus (RABV) in North American bats.
The data set comprises 372 \textit{nucleoprotein} gene sequences from 17 bat species, sampled between 1997 and 2006 across 14 states in the United States \citep{Streicker,Faria2013}.
We estimate RABV ancestral locations and host-jumping history using a Bayesian discrete phylogeographic approach with BSSVS, while simultaneously estimating effective population sizes over time through a Bayesian skygrid coalescent model \citep{gill2012improving}.

Phylogeographic BSSVS inference includes parameters of both integer (number of non-zero transition rates) and categorical (host or location-state) trace types.
In \tracer, a bubble chart visualises the joint probability distribution between two integer or categorical traces (see Figure \ref{fig:4tabs}a).
Circle area is proportional to the joint probability,
with a coloured tile background if this probability reaches a nominal threshold to enhance visibility.
Marginal density plots can also display multiple integer parameters, each with unique colour scales (see Figure \ref{fig:4tabs}b).
With approximately equal numbers of transition rates, both figures suggest similar host and location trait model complexity.
\tracer\ also provides popular visualisations for continuous parameters, including scatter plots for two parameters (see Figure \ref{fig:4tabs}c),
 and extensions for correlations between $\ge\hspace{-0.08em}2$ continuous parameters (Figure \ref{fig:4tabs}d; \citet{Murdoch}).
Colour gradients indicate strength and direction of the correlation, from red (strong negative) to blue (strong positive).
Ellipse shapes re-enforce the strength of correlation, with no correlation appearing as a circle and perfect (anti-)correlation as a line.

\tracer\ reconstructs the demographic history of RABV by drawing the effective population sizes over time (Figure \ref{fig:rabv}).
RABV has successfully established itself in North American bat species, with its effective population size rising steadily throughout recent centuries.
Following a rapid decline at the end of last century, we observe a recent sharp increase in size.

\begin{figure}[t]
\centerline{
\includegraphics[width=0.46\textwidth]{figure_3_rabv-skygrid.pdf}
}
\vspace{-0.25cm}
\caption{Estimating the effective population sizes over time using a Bayesian skygrid demographic reconstruction for rabies virus in North America.}
\label{fig:rabv}
\end{figure}


Other packages are available for the post-processing of MCMC samples.
 `coda' \citep{plummer2006coda} provides some of the functionality of \tracer\ within the R programming environment, while `AWTY' \citep{nylander2007awty} and `RWTY' \citep{warren2017rwty} explore the convergence of the phylogenetic tree parameter itself across multiple MCMC runs.
These alternative packages compute, e.g., Gelman-Rubin diagnostics \citep{gelman1992inference} that \tracer\ currently does not provide.

\vspace{-0.55cm}

\section*{\textsc{Availability}}

\tracer\ is open-source under the GNU lesser general public license and available in both source code (\url{https://github.com/beast-dev/tracer}) and executable (\url{http://beast.community/tracer}) forms.
This latter page also serves up self-contained, step-by-step tutorials covering basic to advanced usage of \tracer\ to summarise posteriors under a variety of phylogenetic models using BEAST and diagnose MCMC chain convergence.
Popular tutorials employ \tracer\ to generate marginal parameter summaries and to infer population dynamics trajectories over time.
\tracer\ requires Java version 1.6 or greater.

\vspace{-0.55cm}

\section*{\textsc{Acknowledgements}}

This work was supported in part by the Wellcome Trust through project 206298/Z/17/Z (Artic Network), the European Union Seventh Framework Programme under grant agreement no. 725422-RESERVOIRDOCS, the National Science Foundation through grant DMS 1264153, the National Institutes of Health under grants R01 AI107034 and U19 AI135995. AJD acknowledges support from Marsden grant contract UOA1611.
GB acknowledges support from the Interne Fondsen KU Leuven / Internal Funds KU Leuven.

\vspace*{-2em}

\bibliographystyle{chicago}
\bibliography{./tracer17}

\end{document}

% COVER
% Please find attached our manuscript introducing the Tracer software tool for publication in Systematic Biology.  While Tracer finds common use amongst phylogenetics and carries several thousand citations via GoogleScholar, there is no publication that announces its availability nor describes its features.  We believe our manuscript will become the de facto citation for Tracer and will quickly emerge as an impactful publication in Systematic Biology.
